libs/x-content/src/main/java/org/elasticsearch/xcontent/DotExpandingXContentParser.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.xcontent;
+
+import java.io.IOException;
+
+/**
+ * An XContentParser that reinterprets a field name containing dots as an object
+ * structure.
+ *
+ * A fieldname named 'foo.bar.baz':... will be parsed instead as 'foo':{'bar':{'baz':...}}
```

Comment:
Maybe want `{@code }` tags for these, just to make the display a little nicer when I mouseover the class.